engine/dlib/src/dlib/android.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 The Defold Foundation
+// Licensed under the Defold License version 1.0 (the "License"); you may not use
+// this file except in compliance with the License.
+//
+// You may obtain a copy of the License, together with FAQs at
+// https://www.defold.com/license
+//
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#if defined(ANDROID)
+
+#include <assert.h>
+#include <dmsdk/dlib/android.h>
+
+extern struct android_app* __attribute__((weak)) g_AndroidApp;
+
+namespace dmAndroid
+{
+
+ThreadAttacher::ThreadAttacher()
+: m_Activity(g_AndroidApp->activity)
+, m_Env(NULL)
+, m_IsAttached(false)
+{
+    JNIEnv* env = 0;
+    if (m_Activity->vm->GetEnv((void **)&m_Env, JNI_VERSION_1_6) != JNI_OK)
+    {
+        m_Activity->vm->AttachCurrentThread(&m_Env, 0);
+        m_IsAttached = true;
+    }
+}
+
+bool ThreadAttacher::Detach()
+{
+    bool ok = true;
+    if (m_IsAttached)
+    {
+        if (m_Env->ExceptionCheck())
+        {
+            m_Env->ExceptionDescribe();
+            ok = false;
+        }
+        m_Env->ExceptionClear();
+        m_Activity->vm->DetachCurrentThread();
+        m_IsAttached = false;
+    }
+    return ok;
+}
```

Comment:
Although the `Detach()`is called automatically, there are also cases when you wish to detach and check there were any errors.